### ISSUE TITLE
Desired workflows for adding new responses

### ISSUE THREAD

**Issue by luna404** [HeRAMS-WHO/herams-backend#1014]
This is a first effort to describe the desired workflow for adding, editing, submitting, and validating responses and is related to the discussion on revising the permissions #1013 

Att the WS level
-	Creating a new draft response
o	Any user with edit permission can create a new draft response and submit it
o	Draft responses are cached locally, only submitted responses should be written to the DB
o	Each user should only be able to cache one draft response per HF
o	A new draft response should always be based on the latest submitted response (same as current behavior)
-	Submitting a response
o	Once a response is submitted it's written to the DB and visible to all users having access to the WS. 
o	A submitted response has passed data validation checks (e.g. must have all mandatory fields completed)
o	It should not be possible to submit multiple responses with the same date of update
o	Upon submitting a response, validators should be informed that a new response was submitted requiring validation
-	Modifying a submitted response
o	Only the user who created the response can edit or delete their response
o	Deleted responses should not appear in the list of responses (as there is no need to view them later they could be hard deleted from the DB)
-	Validating a response
o	Ability to validate should be separate permission from WS owners
o	Validator should be able to view differences between the submitted response and the latest validated response 
o	Validators can either validate or invalidate a response
o	Marking a response as invalid should require justification (i.e. free text field where validators have to explain why it was rejected)
o	Invalidated responses can no longer be edited or deleted
o	Creator of a response should be notified once a response was (in)validated (to be discussed)
o	Editing/deleting of valid responses must be restricted to users with the ability to validate (this needs to be discussed further. While there is a need for some users to edit validated responses, it opens up the risk that validated responses can be overwritten)

If possible, we'd like to be able to track modification of responses(i.e, edits, submission, validaiton, etc.) at the user level, meaning each response should be timestamped and include the name of the user who modified the response.

At the project level (this is for later):
-	Any user with access to the proejct should be able to select a single HFs and download the latest response (i.e. HF reports)
-	Any user with access to the project should be able to submit an anonymous response to a specific HF.
-	Submitted anonymous responses should be visible to all users with permission on the corresponding WS.
-	All users with edit permission to the WS should receive a notification when a new anonymous response was submitted.

**Comment by SamMousa** [comment]
```mermaid
stateDiagram-v2
    [*] --> draft
    draft --> draft: update [C]
    draft --> submitted: submit [C]
    submitted --> draft: revoke [C]
    submitted --> draft: request changes [V]
    submitted --> final: approve [V]
    draft --> [*]: delete [C]
    submitted --> [*]: delete [C]
    final --> draft: update situation [C]
    final --> [*]: delete [A]


```

This is my proposed state diagram for a data response. Any action marked with `[C]` is initiated by a contributor, any action marked with `[V]` is a validator.

I've deviated from some points in the original text:

- Draft responses are submitted to the database.
- Editing of a response is limited to drafts (this prevents validators reviewing something that is being updated), if you want to edit something that's been submitted you have to revoke it, which turns it into a draft, and then update it.
- A finalized response cannot be updated, instead it may only be used as the basis for a new draft.

**Comment by SamMousa** [comment]
@SamuelPetragallo ping.

**Comment by SamuelPetragallo** [comment]
Any point of disagreement left afetr today's meeting?

**Comment by SamMousa** [comment]
No, we should create a diagram for cases where we do not require four-eye-principle. 
A possible solution includes requiring the same flow but just with 1 person, so the same person validates his own data. This is simple from a technical perspective since it follows the same flow.

Advantage is also that it can easily be extended later by changing permissions.
Another pro of this approach is that even without verification you still want to have unpublished drafts.

Next step is to confirm how this interacts with the permissions in #1013, something for tomorrow.
